### ISSUE TITLE
Loading S3 resources through ApplicationContext.getResource broken

### ISSUE THREAD

**Issue by Sintendo** [spring-cloud/spring-cloud-aws#99]
Consider the following example:

```
@Configuration
@PropertySource("classpath:aws.properties")
@EnableContextCredentials(accessKey = "${cloud.aws.credentials.accessKey}", secretKey = "${cloud.aws.credentials.secretKey}")
@EnableContextRegion(region = "${cloud.aws.region.static}")
@EnableContextResourceLoader
public class Application {

    public static void main(String[] args) throws IOException {
        ApplicationContext ctx = new AnnotationConfigApplicationContext(Application.class);
        Resource resource = ctx.getResource("s3://bucket/file.txt");

        System.out.println("Resource class: " + resource.getClass());
        InputStream is = resource.getInputStream();
    }

    @Bean
    public static PropertySourcesPlaceholderConfigurer configurer() {
        return new PropertySourcesPlaceholderConfigurer();
    }

}
```

Executing with 1.0.1.RELEASE prints:

```
Resource class: class org.springframework.cloud.aws.core.io.s3.SimpleStorageResource
```

Executing with 1.0.2.RELEASE or 1.0.3.RELEASE prints:

```
Resource class: class org.springframework.core.io.DefaultResourceLoader$ClassPathContextResource
Exception in thread "main" java.io.FileNotFoundException: class path resource [s3://bucket/file.txt] cannot be opened because it does not exist
    at org.springframework.core.io.ClassPathResource.getInputStream(ClassPathResource.java:172)
    at be.optis.bug.Application.main(Application.java:29)
```

Note that we no longer get a SimpleStorageResource. I'm guessing this is caused by the changes made in commit 3ca473f045f8cd53be6a692258ed4e355a29bc64. Is this new behavior intentional?

**Comment by aemruli** [comment]
Hello @Sintendo indeed the change leads to break your code. We have to do the change as several other places did in the Spring resource loader were broken due to our implementation. With Spring 4.3 (which we will base on our next major release) this issue will be solved (see https://jira.spring.io/browse/SPR-12857). 

Fir timebeing you can get the resource loader from the application context and then load the resource. 

E.g. `application.getBean(ResourceLoader.class);`

I will leave this ticket open in the back log so that we add a dedicated test once migrating to the new resource loader.

**Comment by gonzalad** [comment]
Hello, 

Here's a temporary solution for using s3 resource loading with : 
* Spring Cloud Aws 1.1.3 
* Spring Boot 1.4.1 (and Spring 4.3.3)

```
@Configuration
@Profile(AwsConfig.AWS_PROFILE)
public class AwsConfig { // extends WebMvcConfigurerAdapter {
    public static final String AWS_PROFILE = "aws";
    private static final String S3_PROTOCOL_PREFIX = "s3://";

    @Autowired
    public void registerS3(ConfigurableApplicationContext applicationContext, PathMatchingSimpleStorageResourcePatternResolver s3ResourceLoader) {
        applicationContext.addProtocolResolver(new S3ProtocolResolver(s3ResourceLoader));
    }

    public static class S3ProtocolResolver implements ProtocolResolver {

        private PathMatchingSimpleStorageResourcePatternResolver s3ResourceLoader;

        public S3ProtocolResolver(PathMatchingSimpleStorageResourcePatternResolver s3ResourceLoader) {
            this.s3ResourceLoader = s3ResourceLoader;
        }

        @Override
        public Resource resolve(String location, ResourceLoader resourceLoader) {
            return location.startsWith(S3_PROTOCOL_PREFIX) ? s3ResourceLoader.getResource(location) : null;
        }
    }
}
```
 
It's just a temporary solution since I'm dependending on Spring AWS implementations classes and (at least I think) I cannot use S3 resource loader in PostProcessors.

**Comment by aemruli** [comment]
fixed with 0022146
